compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1484,6 +1486,20 @@ public void visit(BLangBracedOrTupleExpr bracedOrTupleExpr) {
         /* ignore */
     }
 
+    @Override
+    public void visit(BLangScope scopeNode) {
+        visit(scopeNode.scopeBody);
+
+        symbolEnter.defineNode(scopeNode.compensationFunction.function, env);
+        typeChecker.checkExpr(scopeNode.compensationFunction, env);
+        symbolEnter.defineNode(scopeNode, env);
+    }
+
+    @Override
+    public void visit(BLangCompensate node) {
+        /* ignore */
```

Comment:
Scope identifier validation should be done here.

Reply:
This improvement is captured in https://github.com/ballerina-platform/ballerina-lang/issues/9413.